### PR TITLE
WIP for using transceivers

### PR DESCRIPTION
**Pull request by danjenkins**
android not compiling yet... something messed up in the apply of the patch.

The reason we're not discussing things in other Pull requests is a) they're closed and b) the structure of the project changed in v94 and it took quite a bit fo work to match those changes up to the pull requests. The resulting Merge will include the git usernames of the other parties who made it happen so that the right attributions are in place.

thanks to https://github.com/react-native-webrtc/react-native-webrtc/pull/1015 and https://github.com/react-native-webrtc/react-native-webrtc/pull/1014 

Some issues that need looking at include:

* addTransceiver is async but thats against spec
* android doesnt compile yet
* random uuids given to video tracks on ios and so trying to access them can be difficult

This is not a definitive list though

This is meant to be more for discussion around issues; I expect we'll end up with a branch elsewhere

 **important - Branch says 94.1 but it's now based off 98**

### CODE REVIEW

**Comment by saghul** [comment]
Hoping to get to this one very soon. Some extra context: in our testing anything < 96 is very slow to negotiate with transceivers, so I updated WebRTC to M98 (https://github.com/react-native-webrtc/react-native-webrtc/pull/1126) so we can eliminate that from the equation.

File: android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java [hunk 1]
```diff
@@ -82,6 +86,10 @@ public WebRTCModule(ReactApplicationContext reactContext, Options options) {
         mPeerConnectionObservers = new SparseArray<>();
         localStreams = new HashMap<>();
 
+        PeerConnectionFactory.initialize(
+                PeerConnectionFactory.InitializationOptions.builder(reactContext)
+                        .createInitializationOptions());
```
Comment (davidliu) [inline android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java:91]: PeerConnectionFactory is initialized below, so this seems unneeded.

File: android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java [hunk 2]
```diff
@@ -400,6 +424,8 @@ public void peerConnectionInit(ReadableMap configuration, int id) {
         PeerConnection.RTCConfiguration rtcConfiguration = parseRTCConfiguration(configuration);
 
         try {
+            ThreadUtils.runOnExecutor(() -> peerConnectionInitAsync(rtcConfiguration, id));
```
Comment (davidliu) [inline android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java:427]: This seems to duplicate the same code as below, but isn't run synchronously. Seems unneeded.

File: android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java [hunk 3]
```diff
@@ -410,6 +436,17 @@ public void peerConnectionInit(ReadableMap configuration, int id) {
             e.printStackTrace();
             throw new RuntimeException(e);
         }
+
```
Comment (davidliu) [inline android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java:439]: Missing closing brace for `peerConnectionInit()`

File: android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java [hunk 4]
```diff
@@ -777,6 +839,7 @@ public void peerConnectionSetRemoteDescription(ReadableMap sdpMap,
                                                    int id,
                                                    Callback callback) {
         ThreadUtils.runOnExecutor(() -> {
+            peerConnectionAddICECandidateAsync(candidateMap, id, callback);
```
Comment (davidliu) [inline android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java:842]: Unrelated code? Also causes a compile error here, should be removed.

File: android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java [hunk 5]
```diff
@@ -746,6 +807,7 @@ public void onSetSuccess() {
                     WritableMap newSdpMap = Arguments.createMap();
                     newSdpMap.putString("type", newSdp.type.canonicalForm());
                     newSdpMap.putString("sdp", newSdp.description);
+                    newSdpMap.putMap("state", serializeState(id));
```
Comment (davidliu) [inline android/src/main/java/com/oney/WebRTCModule/WebRTCModule.java:810]:
```suggestion
                    newSdpMap.putMap("state", serializeState(pcId));
```
Compile error

File: src/RTCPeerConnection.js [hunk 6]
```diff
@@ -94,13 +97,38 @@ export default class RTCPeerConnection extends defineCustomEventTarget(...PEER_C
...
+
     createOffer(options) {
         return new Promise((resolve, reject) => {
             WebRTCModule.peerConnectionCreateOffer(
                 this._peerConnectionId,
                 RTCUtil.normalizeOfferAnswerOptions(options),
                 (successful, data) => {
                     if (successful) {
+                        this._mergeState(data.state);
                         resolve(data);
```
Comment (davidliu) [inline src/RTCPeerConnection.js:132]:
```suggestion
                        resolve(data.session);
```

Native is now returning a dictionary containing `state` and `session` (the actual sdp). Need to return the `session` value directly to conform to web api spec which returns the `RTCSessionDescription` directly.

File: src/RTCPeerConnection.js [hunk 7]
```diff
@@ -117,6 +145,7 @@ export default class RTCPeerConnection extends defineCustomEventTarget(...PEER_C
...
```
Comment (davidliu) [inline src/RTCPeerConnection.js:149]:
```suggestion
                        resolve(data.session);
```

**Comment by davidliu** [comment]
With the above fixes, I can confirm both iOS/Android working with unified plan and `addTransceivers`.

File: android/src/main/java/com/oney/WebRTCModule/PeerConnectionObserver.java [hunk 8]
```diff
@@ -455,6 +500,15 @@ public void onSignalingChange(PeerConnection.SignalingState signalingState) {
...
         Log.d(TAG, "onAddTrack");
+        if(isUnifiedPlan){
+            MediaStreamTrack track = receiver.track();
+            if(track != null){
+                if(track.kind().equals(MediaStreamTrack.VIDEO_TRACK_KIND)){
+                    videoTrackAdapters.addAdapter(UUID.randomUUID().toString(), (VideoTrack) track);
+                }
+                remoteTracks.put(track.id(), track);
+            }
+        }
```
Comment (murillo128) [inline android/src/main/java/com/oney/WebRTCModule/PeerConnectionObserver.java:511]: Shouldn't we fire an event here?

**Comment by saghul** [comment]
After spending a couple of days tinkering with this PR I am going to abandon ship. The RtpSender / receiver implementation makes it alas unusable. There are also a bunch of unnecessary and / or incorrect changes or weird merges.

Here is what I'm currently working on, PR coming soon:

- Drop plan B altogether
- Add senders / receivers / addTrack / ontrack
- Add transceivers

I think the only way forward it to pull a Safari and switch plans abruptly.

With this plan (pun intended) I would be able to test the core with Jitsi Meet at the middle point, and add transceivers just after.

**Comment by murillo128** [comment]
> I'll open a new one shortly and we can see how to join forces there, WDYT?

Perfect!
